classes/submissionFile/DAO.php

```diff
@@ -67,19 +76,31 @@ public function newDataObject(): SubmissionFile
     }
 
     /**
-     * @copydoc EntityDAO::get()
+     * Get an submission file by its ID, and optionaly by its submission ID
      */
-    public function get(int $id): ?SubmissionFile
+    public function get(int $id, int $submissionId = null): ?SubmissionFile
     {
         $query = new Collector($this);
         $row = $query
             ->getQueryBuilder()
             ->where($this->primaryKeyColumn, '=', $id)
+            ->when($submissionId !== null, fn (Builder $query, int $submissionId) => $query->where('sf.submission_id', $submissionId))
             ->first();
 
         return $row ? $this->fromRow($row) : null;
     }
 
+    /**
+     * Check if a submission file exists with this ID, and optional submission ID
+     */
+    public function exists(int $id, int $submissionId = null): bool
+    {
+        return DB::table($this->table)
+            ->where($this->primaryKeyColumn, '=', $id)
+            ->when($submissionId !== null, fn (Builder $query, int $submissionId) => $query->where($this->getParentColumn(), $submissionId))
+            ->exists();
+    }
```

Comment:
Shouldn't these methods be implemented by the trait?

Reply:
submissionFile, author and user need a different get() function -- they need joins with other tables. Thus, the trait get() method cannot be used in these DAOs. Thus, the get() method needs to be implemented in these DAOs. That's why we cannot use the trait at all -- because it contains the same function name get().
The trait exists() method could be used by these DAOs, but because we cannot use the trait, also the exists() would need to be implemented in the DAOs.

Reply:
maybe to separate get() and exists() into different traits: EntityGetWithParent and EntityExistsWithParent?

Reply:
No I think it's fine to keep it simple. I didn't see the difference in `submissionFile/DAO::get()` but I see now, you need the table alias, `sf.`. :+1: